sdk/src/Core/Amazon.Runtime/Internal/Util/StreamingUtf8JsonReader.cs

```diff
@@ -0,0 +1,158 @@
+        private static JsonReaderOptions _jsonOptions = new JsonReaderOptions
+        {
+            AllowTrailingCommas = true
+        };
+        /// <summary>
+        /// The UTF8JsonReader attached to the instance. 
+        /// </summary>
+        public Utf8JsonReader Reader
+        {
+            get
+            {
+                return _reader;
+            }
+        }
+
+        private Stream _stream;
+        private byte[] _buffer;
+        public StreamingUtf8JsonReader(Stream stream)
+        {
+            if (stream is null)
+                throw new ArgumentException("Stream must not be null. Please initialize a stream and pass it into the constructor.");
+
+            _stream = stream;
+            _buffer = ArrayPool<byte>.Shared.Rent(AWSConfigs.StreamingUtf8JsonReaderBufferSize ?? 4096);
+            // need to initialize the reader even if the buffer is empty because auto-default of unassigned fields is only 
+            // supported in C# 11+
+
+            int utf8BomLength = JsonConstants.Utf8Bom.Length;
+            Debug.Assert(_buffer.Length >= utf8BomLength);
+
+            int bytesRead = _stream.Read(_buffer, 0, _buffer.Length);
+            int start = 0;
+            if (_buffer.AsSpan().StartsWith(JsonConstants.Utf8Bom))
+            {
+                start += utf8BomLength;
+                bytesRead -= utf8BomLength;
+            }
+
+            _reader = new Utf8JsonReader(_buffer.AsSpan(start, bytesRead), isFinalBlock: bytesRead == 0, new JsonReaderState(_jsonOptions));
+        }
+
+        /// <summary>
+        /// Reads data from the buffer and fetches more data from the stream to fill the buffer it more data exists.
+        /// </summary>
+        /// <returns>true if there is more data, false otherwise.</returns>
+        public bool Read()
+        {
+            // hasMoreData can return false if the value starts in one buffer and leaks into the next buffer
+            bool hasMoreData = _reader.Read();
+
+            while (!hasMoreData)
+            {
+                GetMoreBytesFromStream(_stream, ref _buffer, ref _reader);
+
+                if (_reader.IsFinalBlock)
+                    break;
+
+                hasMoreData = _reader.Read();
+            }
+
+            return hasMoreData;
+        }
+
+        private static void GetMoreBytesFromStream(Stream stream, ref byte[] buffer, ref Utf8JsonReader reader)
+        {
+            int bytesRead = 0;
+            int previousBufferLength = buffer.Length;
+            if (reader.BytesConsumed < buffer.Length)
+            {
+                ReadOnlySpan<byte> leftover = buffer.AsSpan().Slice((int)reader.BytesConsumed);
+                // If BytesConsumed is 0 that means that the previous Read failed because the JSON token was too large to fit in the buffer.
+                // In that case we need to resize the buffer and try again to read the JSON token.
+                if (reader.BytesConsumed == 0)
```

Comment:
If I'm misunderstanding the code in here with the `leftover` and `previousBufferLength` variables that is why this code needs some code. I can read the code but it is hard to understand the whys. That was why in my sample I gave you I broke the logic into spaced logic with comments for that blocks so you could understand the stages of the logic.

Reply:
Gotcha, I've added much more comments